{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3a7fe081-a496-4898-8160-c93151d4535d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "# Get the path to the validate_assay folder\n",
    "module_path = os.path.abspath(os.path.join('..', 'validate_assay'))\n",
    "\n",
    "# Add the folder to the system path\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "792feb6c-11b4-450e-a109-83128a4ab38b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pcrvalidationtools import *\n",
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "779ae401-dabc-4ecc-afb1-7f36a377da0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# files \n",
    "INPUT_PATH = \"../Data/v4_08262024/\"\n",
    "datasets_file = INPUT_PATH + 'enterovirus_metadata.tsv'\n",
    "clustered_file =  INPUT_PATH + 'ev_clustered100.fasta'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c937696d-84f8-4880-8a13-1da727068333",
   "metadata": {},
   "source": [
    "# 1. Enterovirus "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbc2c3a6-11ac-4d2d-a2c3-da5de03b896c",
   "metadata": {},
   "source": [
    "## 1a. Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2baf0176-0bec-4e07-8739-c9ef3ebcdacd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 6495 sequences after filtering for complete human-host sequences with a collection date\n"
     ]
    }
   ],
   "source": [
    "# filter for entered collection date, human host, completeness \n",
    "df_filter1 = filter_data(datasets_file)\n",
    "\n",
    "print(f'There are {df_filter1.shape[0]} sequences after filtering for complete human-host sequences with a collection date')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4c4e838b-8ade-4b63-9a51-e8cd28cdd669",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered accessions saved to enterovirus_filtered_acc.txt\n"
     ]
    }
   ],
   "source": [
    "# create a text file of one accessions per line to retrieve sequences for\n",
    "accession_output_filename = \"enterovirus_filtered_acc.txt\"\n",
    "df_filter1['Accession'].to_csv(INPUT_PATH + accession_output_filename, index=False, header=False)\n",
    "\n",
    "print(f'Filtered accessions saved to {accession_output_filename}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "26155666-e1af-4f6d-8bd5-ab12b17799eb",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 1b. Cluster at 100% identity with CD-HIT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "61e32aef-1993-4b51-ae0a-194cd9574ca2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# return list of accessions from fasta headers\n",
    "# ex. header: >PP191126.1 Coxsackievirus A6 strain CVA6/16/GJ/KOR/2022, complete genome \n",
    "def get_acc_fasta(file_path):\n",
    "    # parse headers\n",
    "    headers = []\n",
    "    with open(file_path, 'r') as fasta_file:\n",
    "        for line in fasta_file:\n",
    "            if line.startswith('>'):\n",
    "                header = line.strip()[1:]\n",
    "                headers.append(header)\n",
    "    # retrieve accession from header\n",
    "    accessions = [h.split(' ')[0] for h in headers]\n",
    "    return accessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23d7c370-6022-48c6-85a3-a409954a6fcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 6149 sequences after dereplication\n",
      "346 sequences were removed here\n"
     ]
    }
   ],
   "source": [
    "# read in cd-hit output \n",
    "dereplicated_acc = get_acc_fasta(clustered_file)\n",
    "\n",
    "# remove duplicate sequences \n",
    "df_filter2 = df_filter1[df_filter1['Accession'].isin(dereplicated_acc)].copy()\n",
    "\n",
    "print(f'There are {df_filter2.shape[0]} sequences after dereplication')\n",
    "print(f'{df_filter1.shape[0] - df_filter2.shape[0]} sequences were removed here')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da6e6a9e-5cec-40dc-a18b-f38ff8b828e3",
   "metadata": {},
   "source": [
    "## 1c. Filter for genomes with 5' UTRs greater than 400 nt long"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "8856c8f1-3f1c-4458-87e8-656057b5b1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use entrez to determine the 5' utr length \n",
    "\n",
    "# info for NCBI Entrez\n",
    "Entrez.email = # email here\n",
    "Entrez.api_key = # API key here \n",
    "\n",
    "# returns a list of queries\n",
    "# each query is a string of comma separated accessions\n",
    "# this breaks up the requests we send to NCBI\n",
    "# accessions (list): list of accession number strings\n",
    "def define_query(identifiers):\n",
    "    # if getting invalid uid error, try lowering batch size\n",
    "    BATCH_SIZE = 1000\n",
    "    queries = []\n",
    "    size = len(identifiers)\n",
    "    iters = (size // BATCH_SIZE) + (size % BATCH_SIZE > 0)\n",
    "    \n",
    "    # make sure the list are strings\n",
    "    identifiers = list(map(str, identifiers))\n",
    "    \n",
    "    for i in range(iters):\n",
    "        start_range = i * BATCH_SIZE\n",
    "        end_range = start_range + BATCH_SIZE\n",
    "        query_str = \", \".join(identifiers[start_range:end_range])\n",
    "        queries.append(query_str)\n",
    "    return queries\n",
    "\n",
    "# returns genbank records for the list of accessions\n",
    "# accessions : list of accession strings\n",
    "def get_gb_records(accessions):\n",
    "    gb_records = []\n",
    "    queries = define_query(accessions)\n",
    "    for query in queries:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=query, rettype=\"gb\", retmode=\"text\")\n",
    "        gb_records.extend(SeqIO.parse(handle, \"genbank\"))\n",
    "        handle.close()\n",
    "    return gb_records\n",
    "\n",
    "# returns length of 5' utr \n",
    "# record: gb record\n",
    "def get_utr_len(record):\n",
    "    utr_len = np.nan\n",
    "    # parse through features for CDS or 5'UTR\n",
    "    for feature in record.features:\n",
    "        if feature.type == '5\\'UTR':\n",
    "            # get 5' utr location\n",
    "            loc = feature.location\n",
    "            # define utr length as start of cds annotation\n",
    "            utr_len = loc.start\n",
    "            # if direction is reverse, calculate utr length\n",
    "            if loc.strand == -1:\n",
    "                utr_len = len(record.seq) - loc.start\n",
    "            break\n",
    "        if feature.type == 'CDS':\n",
    "            # get cds location\n",
    "            loc = feature.location\n",
    "            # define utr length as start of cds annotation\n",
    "            utr_len = loc.start\n",
    "            # if direction is reverse, calculate utr length\n",
    "            if loc.strand == -1:\n",
    "                utr_len = len(record.seq) - loc.end\n",
    "            break\n",
    "    return utr_len\n",
    "\n",
    "# returns a df with accession number and corresponding 5' utr length\n",
    "def accession_utr_df(gb_records):\n",
    "    # parse records and get utr length record\n",
    "    df_dict = {'Accession': [], '5\\'UTR Length': []}\n",
    "    for record in gb_records:\n",
    "        df_dict['Accession'].append(record.id)\n",
    "        df_dict['5\\'UTR Length'].append(get_utr_len(record))\n",
    "\n",
    "    # return df\n",
    "    return pd.DataFrame(df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0ab58337-f84b-4a01-afd1-338aca15388a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# retrieve length of 5' utr for filtered genomes\n",
    "gb_records = get_gb_records(df_filter2['Accession'].tolist())\n",
    "utr_df = accession_utr_df(gb_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "44d340f2-e207-4fca-9960-f738e77f19c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>5'UTR Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR204681.1</td>\n",
       "      <td>694.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR204682.1</td>\n",
       "      <td>709.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR220808.1</td>\n",
       "      <td>572.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR220809.1</td>\n",
       "      <td>563.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR230416.1</td>\n",
       "      <td>710.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6144</th>\n",
       "      <td>AB779614.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6145</th>\n",
       "      <td>AB779615.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6146</th>\n",
       "      <td>AB779616.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6147</th>\n",
       "      <td>AB779617.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6148</th>\n",
       "      <td>AB779618.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6149 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession  5'UTR Length\n",
       "0     OR204681.1         694.0\n",
       "1     OR204682.1         709.0\n",
       "2     OR220808.1         572.0\n",
       "3     OR220809.1         563.0\n",
       "4     OR230416.1         710.0\n",
       "...          ...           ...\n",
       "6144  AB779614.1           0.0\n",
       "6145  AB779615.1           0.0\n",
       "6146  AB779616.1           0.0\n",
       "6147  AB779617.1           0.0\n",
       "6148  AB779618.1           0.0\n",
       "\n",
       "[6149 rows x 2 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ef2be325-fd65-472b-b69d-998703fbd935",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 4676 sequences after removing 5' UTRs < 400nt\n",
      "1473 sequences were removed here\n"
     ]
    }
   ],
   "source": [
    "# filter genomes for 5'UTR length >= 400 \n",
    "valid_utr_acc = utr_df[utr_df['5\\'UTR Length'] >= 400]['Accession'].tolist()\n",
    "df_filter3 = df_filter2[df_filter2['Accession'].isin(valid_utr_acc)].copy()\n",
    "\n",
    "print(f'There are {df_filter3.shape[0]} sequences after removing 5\\' UTRs < 400nt')\n",
    "print(f'{df_filter2.shape[0] - df_filter3.shape[0]} sequences were removed here')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d5593197-ad13-45b2-bf7a-1c76eb9ae851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Completely filtered genomes saved to ev_metadata_filtered.tsv\n"
     ]
    }
   ],
   "source": [
    "# save filtered datasets as tsv file \n",
    "output_filename = 'ev_metadata_filtered.tsv'\n",
    "df_filter3.to_csv(INPUT_PATH + output_filename, sep='\\t', index=False)\n",
    "\n",
    "print(f'Completely filtered genomes saved to {output_filename}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "023e422f-b1c4-4e11-85bc-cb549c4b22b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Virus Name</th>\n",
       "      <th>Virus Taxonomic ID</th>\n",
       "      <th>Host Taxonomic ID</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Isolate Collection date</th>\n",
       "      <th>Release date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>OR204681.1</td>\n",
       "      <td>Coxsackievirus B3</td>\n",
       "      <td>12072</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2021</td>\n",
       "      <td>2023-08-09T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>OR204682.1</td>\n",
       "      <td>Coxsackievirus B3</td>\n",
       "      <td>12072</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2021</td>\n",
       "      <td>2023-08-09T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>OR220808.1</td>\n",
       "      <td>Rhinovirus C</td>\n",
       "      <td>463676</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2023-02-23</td>\n",
       "      <td>2023-09-13T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>OR220809.1</td>\n",
       "      <td>Rhinovirus A</td>\n",
       "      <td>147711</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2023-02-24</td>\n",
       "      <td>2023-09-13T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>OR230416.1</td>\n",
       "      <td>enterovirus D68</td>\n",
       "      <td>42789</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2020</td>\n",
       "      <td>2023-07-24T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4671</th>\n",
       "      <td>AB769160.1</td>\n",
       "      <td>Coxsackievirus A24</td>\n",
       "      <td>12089</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-06</td>\n",
       "      <td>2012-12-05T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4672</th>\n",
       "      <td>AB769162.2</td>\n",
       "      <td>Coxsackievirus A24</td>\n",
       "      <td>12089</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-06</td>\n",
       "      <td>2012-12-05T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4673</th>\n",
       "      <td>AB769163.1</td>\n",
       "      <td>Coxsackievirus A24</td>\n",
       "      <td>12089</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-10</td>\n",
       "      <td>2012-12-05T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4674</th>\n",
       "      <td>AB769164.2</td>\n",
       "      <td>Coxsackievirus A24</td>\n",
       "      <td>12089</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-10</td>\n",
       "      <td>2012-12-05T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4675</th>\n",
       "      <td>AB769165.1</td>\n",
       "      <td>Coxsackievirus A24</td>\n",
       "      <td>12089</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-10</td>\n",
       "      <td>2012-12-05T00:00:00Z</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4676 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession          Virus Name  Virus Taxonomic ID  Host Taxonomic ID  \\\n",
       "0     OR204681.1   Coxsackievirus B3               12072             9606.0   \n",
       "1     OR204682.1   Coxsackievirus B3               12072             9606.0   \n",
       "2     OR220808.1        Rhinovirus C              463676             9606.0   \n",
       "3     OR220809.1        Rhinovirus A              147711             9606.0   \n",
       "4     OR230416.1     enterovirus D68               42789             9606.0   \n",
       "...          ...                 ...                 ...                ...   \n",
       "4671  AB769160.1  Coxsackievirus A24               12089             9606.0   \n",
       "4672  AB769162.2  Coxsackievirus A24               12089             9606.0   \n",
       "4673  AB769163.1  Coxsackievirus A24               12089             9606.0   \n",
       "4674  AB769164.2  Coxsackievirus A24               12089             9606.0   \n",
       "4675  AB769165.1  Coxsackievirus A24               12089             9606.0   \n",
       "\n",
       "     Completeness Isolate Collection date          Release date  \n",
       "0        COMPLETE                    2021  2023-08-09T00:00:00Z  \n",
       "1        COMPLETE                    2021  2023-08-09T00:00:00Z  \n",
       "2        COMPLETE              2023-02-23  2023-09-13T00:00:00Z  \n",
       "3        COMPLETE              2023-02-24  2023-09-13T00:00:00Z  \n",
       "4        COMPLETE                    2020  2023-07-24T00:00:00Z  \n",
       "...           ...                     ...                   ...  \n",
       "4671     COMPLETE                 2011-06  2012-12-05T00:00:00Z  \n",
       "4672     COMPLETE                 2011-06  2012-12-05T00:00:00Z  \n",
       "4673     COMPLETE                 2011-10  2012-12-05T00:00:00Z  \n",
       "4674     COMPLETE                 2011-10  2012-12-05T00:00:00Z  \n",
       "4675     COMPLETE                 2011-10  2012-12-05T00:00:00Z  \n",
       "\n",
       "[4676 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check file saved \n",
    "ev_data = pd.read_csv(INPUT_PATH+output_filename, sep='\\t') \n",
    "ev_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "758bf237-2d9d-4b75-bff0-dc289a260bbc",
   "metadata": {},
   "source": [
    "## 1d. Species data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "66481bd5-a1b8-46f4-9d9e-ba05f65f4d27",
   "metadata": {},
   "outputs": [],
   "source": [
    "# returns genbank records for the list of accessions\n",
    "# accessions : list of accession strings\n",
    "def get_taxonomy_records(taxids):\n",
    "    taxonomy_records = []\n",
    "    queries = define_query(taxids)\n",
    "    for query in queries:\n",
    "        handle = Entrez.efetch(db=\"taxonomy\", id=query, retmode=\"xml\")\n",
    "        records = Entrez.read(handle)\n",
    "        handle.close()\n",
    "        taxonomy_records.extend(records)\n",
    "    return taxonomy_records\n",
    "\n",
    "# returns a df with taxid and corresponding collapse taxid \n",
    "def taxonomy_collapse_df(records, target_taxid):\n",
    "    # parse records and determine collapse taxid \n",
    "    df_dict = {'Virus Taxonomic ID': [], 'Collapse TaxId': [], 'Collapse Name': []}\n",
    "    for record in records:\n",
    "        \n",
    "        taxid = int(record['TaxId'])\n",
    "        collapse_taxid = None\n",
    "        collapse_name = None\n",
    "        \n",
    "        # ordered from top of lineage to bottom (most specific)\n",
    "        taxonomy_lineage = [node['TaxId'] for node in record['LineageEx']]\n",
    "        \n",
    "        # find the collapse taxid\n",
    "        if str(target_taxid) in taxonomy_lineage:\n",
    "            # the taxonomy level we want to collapse at is one level below the target \n",
    "            collapse_index = taxonomy_lineage.index(str(target_taxid)) + 1 \n",
    "            # if index within range\n",
    "            if collapse_index < len(taxonomy_lineage):\n",
    "                collapse_taxid = taxonomy_lineage[collapse_index] \n",
    "                collapse_name = record['LineageEx'][collapse_index]['ScientificName']\n",
    "            else:\n",
    "                collapse_taxid = taxid \n",
    "                collapse_name = record['ScientificName']\n",
    "                \n",
    "        df_dict['Virus Taxonomic ID'].append(int(taxid))\n",
    "        df_dict['Collapse TaxId'].append(int(collapse_taxid))\n",
    "        df_dict['Collapse Name'].append(collapse_name)\n",
    "\n",
    "    # return df\n",
    "    return pd.DataFrame(df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7460c37c-2b53-4734-bde4-b8ddc878e798",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Virus Taxonomic ID</th>\n",
       "      <th>Collapse TaxId</th>\n",
       "      <th>Collapse Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1435148</td>\n",
       "      <td>138948</td>\n",
       "      <td>Enterovirus A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>185889</td>\n",
       "      <td>147712</td>\n",
       "      <td>Rhinovirus B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>185891</td>\n",
       "      <td>147711</td>\n",
       "      <td>Rhinovirus A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2094116</td>\n",
       "      <td>463676</td>\n",
       "      <td>Rhinovirus C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>185893</td>\n",
       "      <td>147711</td>\n",
       "      <td>Rhinovirus A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>166</th>\n",
       "      <td>35293</td>\n",
       "      <td>138949</td>\n",
       "      <td>Enterovirus B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>35295</td>\n",
       "      <td>138949</td>\n",
       "      <td>Enterovirus B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>35294</td>\n",
       "      <td>138949</td>\n",
       "      <td>Enterovirus B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>169</th>\n",
       "      <td>297248</td>\n",
       "      <td>138948</td>\n",
       "      <td>Enterovirus A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>170</th>\n",
       "      <td>1141754</td>\n",
       "      <td>138951</td>\n",
       "      <td>Enterovirus D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>171 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Virus Taxonomic ID  Collapse TaxId  Collapse Name\n",
       "0               1435148          138948  Enterovirus A\n",
       "1                185889          147712   Rhinovirus B\n",
       "2                185891          147711   Rhinovirus A\n",
       "3               2094116          463676   Rhinovirus C\n",
       "4                185893          147711   Rhinovirus A\n",
       "..                  ...             ...            ...\n",
       "166               35293          138949  Enterovirus B\n",
       "167               35295          138949  Enterovirus B\n",
       "168               35294          138949  Enterovirus B\n",
       "169              297248          138948  Enterovirus A\n",
       "170             1141754          138951  Enterovirus D\n",
       "\n",
       "[171 rows x 3 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ev_txid = 12059\n",
    "taxids = set(ev_data['Virus Taxonomic ID'])\n",
    "records = get_taxonomy_records(taxids)\n",
    "taxonomy_collapse = taxonomy_collapse_df(records, ev_txid)\n",
    "taxonomy_collapse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ad2ff7c5-9517-49f7-8c67-bd8974bfda07",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saved to ev_taxonomy_collapse.tsv\n"
     ]
    }
   ],
   "source": [
    "# save as tsv file \n",
    "output_filename = 'ev_taxonomy_collapse.tsv'\n",
    "taxonomy_collapse.to_csv(INPUT_PATH + output_filename, sep='\\t', index=False)\n",
    "\n",
    "print(f'Saved to {output_filename}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
