{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3a7fe081-a496-4898-8160-c93151d4535d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "# Get the path to the validate_assay folder\n",
    "module_path = os.path.abspath(os.path.join('..', 'validate_assay'))\n",
    "\n",
    "# Add the folder to the system path\n",
    "if module_path not in sys.path:\n",
    "    sys.path.append(module_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "792feb6c-11b4-450e-a109-83128a4ab38b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pcrvalidationtools import *\n",
    "from Bio import Entrez, SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "779ae401-dabc-4ecc-afb1-7f36a377da0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# files \n",
    "INPUT_PATH = \"../Data/\"\n",
    "datasets_file = INPUT_PATH + 'enterovirus_metadata.tsv'\n",
    "clustered_file =  INPUT_PATH + 'ev_clustered100.fasta'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2baf0176-0bec-4e07-8739-c9ef3ebcdacd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 6488 sequences after filtering for complete human-host sequences with a collection date\n"
     ]
    }
   ],
   "source": [
    "# filter for entered collection date, human host, completeness \n",
    "df_filter1 = filter_data(datasets_file)\n",
    "\n",
    "print(f'There are {df_filter1.shape[0]} sequences after filtering for complete human-host sequences with a collection date')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4c4e838b-8ade-4b63-9a51-e8cd28cdd669",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Filtered accessions saved to enterovirus_filtered_acc.txt\n"
     ]
    }
   ],
   "source": [
    "# create a text file of one accessions per line to retrieve sequences for\n",
    "accession_output_filename = \"enterovirus_filtered_acc.txt\"\n",
    "df_filter1['Accession'].to_csv(INPUT_PATH + accession_output_filename, index=False, header=False)\n",
    "\n",
    "print(f'Filtered accessions saved to {accession_output_filename}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "8a00a835-8419-413b-8592-02f7e9479dab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cluster at 100% identity \n",
    "\n",
    "# get list of accessions from headers of fasta file\n",
    "# ex. header: >PP191126.1 Coxsackievirus A6 strain CVA6/16/GJ/KOR/2022, complete genome \n",
    "def get_acc_fasta(file_path):\n",
    "    # parse headers\n",
    "    headers = []\n",
    "    with open(file_path, 'r') as fasta_file:\n",
    "        for line in fasta_file:\n",
    "            if line.startswith('>'):\n",
    "                header = line.strip()[1:]\n",
    "                headers.append(header)\n",
    "    # retrieve accession from header\n",
    "    accessions = [h.split(' ')[0] for h in headers]\n",
    "    return accessions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f5e30a5b-9797-48e7-9149-d16a75b9d4af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 5974 sequences after dereplication\n",
      "514 sequences were removed here\n"
     ]
    }
   ],
   "source": [
    "# retrieve accessions after 100% dereplication\n",
    "dereplicated_acc = get_acc_fasta(clustered_file)\n",
    "\n",
    "# remove replicated sequences \n",
    "df_filter2 = df_filter1[df_filter1['Accession'].isin(dereplicated_acc)].copy()\n",
    "\n",
    "print(f'There are {df_filter2.shape[0]} sequences after dereplication')\n",
    "print(f'{df_filter1.shape[0] - df_filter2.shape[0]} sequences were removed here')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "8856c8f1-3f1c-4458-87e8-656057b5b1b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use entrez to determine the 5' utr length \n",
    "\n",
    "# info for NCBI Entrez\n",
    "Entrez.email = \"vantrinh@berkeley.edu\"\n",
    "Entrez.api_key = '621bd974d7f13581ac5112cf933598a06b08'\n",
    "\n",
    "# returns a list of queries\n",
    "# each query is a string of comma separated accessions\n",
    "# this breaks up the requests we send to NCBI\n",
    "# accessions (list): list of accession number strings\n",
    "def define_query(accessions):\n",
    "    # if getting invalid uid error, try lowering batch size\n",
    "    BATCH_SIZE = 1000\n",
    "    queries = []\n",
    "    size = len(accessions)\n",
    "    iters = (size // BATCH_SIZE) + (size % BATCH_SIZE > 0)\n",
    "    for i in range(iters):\n",
    "        start_range = i * BATCH_SIZE\n",
    "        end_range = start_range + BATCH_SIZE\n",
    "        accession_str = \", \".join(accessions[start_range:end_range])\n",
    "        queries.append(accession_str)\n",
    "    return queries\n",
    "\n",
    "# returns genbank records for the list of accessions\n",
    "# accessions : list of accession strings\n",
    "def get_gb_records(accessions):\n",
    "    gb_records = []\n",
    "    queries = define_query(accessions)\n",
    "    for query in queries:\n",
    "        handle = Entrez.efetch(db=\"nucleotide\", id=query, rettype=\"gb\", retmode=\"text\")\n",
    "        gb_records.extend(SeqIO.parse(handle, \"genbank\"))\n",
    "        handle.close()\n",
    "    return gb_records\n",
    "\n",
    "# returns length of 5' utr \n",
    "# record: gb record\n",
    "def get_utr_len(record):\n",
    "    utr_len = np.nan\n",
    "    # parse through features for CDS or 5'UTR\n",
    "    for feature in record.features:\n",
    "        if feature.type == '5\\'UTR':\n",
    "            # get 5' utr location\n",
    "            loc = feature.location\n",
    "            # define utr length as start of cds annotation\n",
    "            utr_len = loc.start\n",
    "            # if direction is reverse, calculate utr length\n",
    "            if loc.strand == -1:\n",
    "                utr_len = len(record.seq) - loc.start\n",
    "            break\n",
    "        if feature.type == 'CDS':\n",
    "            # get cds location\n",
    "            loc = feature.location\n",
    "            # define utr length as start of cds annotation\n",
    "            utr_len = loc.start\n",
    "            # if direction is reverse, calculate utr length\n",
    "            if loc.strand == -1:\n",
    "                utr_len = len(record.seq) - loc.end\n",
    "            break\n",
    "    return utr_len\n",
    "\n",
    "# returns a df with accession number and corresponding 5' utr length\n",
    "def accession_utr_df(gb_records):\n",
    "    # parse records and get utr length record\n",
    "    df_dict = {'Accession': [], '5\\'UTR Length': []}\n",
    "    for record in gb_records:\n",
    "        df_dict['Accession'].append(record.id)\n",
    "        df_dict['5\\'UTR Length'].append(get_utr_len(record))\n",
    "\n",
    "    # return df\n",
    "    return pd.DataFrame(df_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "0ab58337-f84b-4a01-afd1-338aca15388a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/vantrinh/anaconda3/lib/python3.10/site-packages/Bio/GenBank/Scanner.py:1791: BiopythonParserWarning: Structured comment not parsed for MH484166. Is it malformed?\n",
      "  warnings.warn(\n",
      "/Users/vantrinh/anaconda3/lib/python3.10/site-packages/Bio/GenBank/Scanner.py:1791: BiopythonParserWarning: Structured comment not parsed for MH484165. Is it malformed?\n",
      "  warnings.warn(\n",
      "/Users/vantrinh/anaconda3/lib/python3.10/site-packages/Bio/GenBank/Scanner.py:1791: BiopythonParserWarning: Structured comment not parsed for MH484164. Is it malformed?\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# retrieve length of 5' utr for filtered genomes\n",
    "gb_records = get_gb_records(df_filter2['Accession'].tolist())\n",
    "utr_df = accession_utr_df(gb_records)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "44d340f2-e207-4fca-9960-f738e77f19c4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>5'UTR Length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP191126.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP191125.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP191124.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP191123.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP191122.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5969</th>\n",
       "      <td>AB705310.1</td>\n",
       "      <td>742.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5970</th>\n",
       "      <td>AB705309.1</td>\n",
       "      <td>742.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5971</th>\n",
       "      <td>AB705308.1</td>\n",
       "      <td>742.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5972</th>\n",
       "      <td>AB686524.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5973</th>\n",
       "      <td>AB678778.1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5974 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession  5'UTR Length\n",
       "0     PP191126.1           0.0\n",
       "1     PP191125.1           0.0\n",
       "2     PP191124.1           0.0\n",
       "3     PP191123.1           0.0\n",
       "4     PP191122.1           0.0\n",
       "...          ...           ...\n",
       "5969  AB705310.1         742.0\n",
       "5970  AB705309.1         742.0\n",
       "5971  AB705308.1         742.0\n",
       "5972  AB686524.1           0.0\n",
       "5973  AB678778.1           0.0\n",
       "\n",
       "[5974 rows x 2 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "utr_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "ef2be325-fd65-472b-b69d-998703fbd935",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 4555 sequences after removing 5' UTRs < 400nt\n",
      "1419 sequences were removed here\n"
     ]
    }
   ],
   "source": [
    "# filter genomes for 5'UTR length >= 400 \n",
    "valid_utr_acc = utr_df[utr_df['5\\'UTR Length'] >= 400]['Accession'].tolist()\n",
    "df_filter3 = df_filter2[df_filter2['Accession'].isin(valid_utr_acc)].copy()\n",
    "\n",
    "print(f'There are {df_filter3.shape[0]} sequences after removing 5\\' UTRs < 400nt')\n",
    "print(f'{df_filter2.shape[0] - df_filter3.shape[0]} sequences were removed here')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d5593197-ad13-45b2-bf7a-1c76eb9ae851",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Completely filtered genomes saved to ev_datasets_filtered.tsv\n"
     ]
    }
   ],
   "source": [
    "# save filtered datasets as tsv file \n",
    "output_filename = 'ev_datasets_filtered.tsv'\n",
    "df_filter3.to_csv(INPUT_PATH + output_filename, sep='\\t', index=False)\n",
    "\n",
    "print(f'Completely filtered genomes saved to {output_filename}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fe6b7951-2283-41c6-a2af-d8f82472b4d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Virus Name</th>\n",
       "      <th>Virus Taxonomic ID</th>\n",
       "      <th>Host Taxonomic ID</th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Isolate Collection date</th>\n",
       "      <th>Release date</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PP157596.1</td>\n",
       "      <td>enterovirus D68</td>\n",
       "      <td>42789</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2020</td>\n",
       "      <td>2024-01-24T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PP157595.1</td>\n",
       "      <td>enterovirus D68</td>\n",
       "      <td>42789</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2020</td>\n",
       "      <td>2024-01-24T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PP125279.1</td>\n",
       "      <td>Rhinovirus B</td>\n",
       "      <td>147712</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2021-06-07</td>\n",
       "      <td>2024-01-24T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PP125278.1</td>\n",
       "      <td>Rhinovirus B</td>\n",
       "      <td>147712</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2021-05-13</td>\n",
       "      <td>2024-01-24T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PP028461.1</td>\n",
       "      <td>Coxsackievirus A24</td>\n",
       "      <td>12089</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2023-09</td>\n",
       "      <td>2024-01-11T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4550</th>\n",
       "      <td>AB769152.1</td>\n",
       "      <td>Coxsackievirus A24</td>\n",
       "      <td>12089</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-06</td>\n",
       "      <td>2012-12-05T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4551</th>\n",
       "      <td>AB705311.1</td>\n",
       "      <td>Echovirus E6</td>\n",
       "      <td>12062</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-09-09</td>\n",
       "      <td>2012-09-14T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4552</th>\n",
       "      <td>AB705310.1</td>\n",
       "      <td>Echovirus E6</td>\n",
       "      <td>12062</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-09-07</td>\n",
       "      <td>2012-09-14T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4553</th>\n",
       "      <td>AB705309.1</td>\n",
       "      <td>Echovirus E6</td>\n",
       "      <td>12062</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-09-02</td>\n",
       "      <td>2012-09-14T00:00:00Z</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4554</th>\n",
       "      <td>AB705308.1</td>\n",
       "      <td>Echovirus E6</td>\n",
       "      <td>12062</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>COMPLETE</td>\n",
       "      <td>2011-08-12</td>\n",
       "      <td>2012-09-14T00:00:00Z</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4555 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession          Virus Name  Virus Taxonomic ID  Host Taxonomic ID  \\\n",
       "0     PP157596.1     enterovirus D68               42789             9606.0   \n",
       "1     PP157595.1     enterovirus D68               42789             9606.0   \n",
       "2     PP125279.1        Rhinovirus B              147712             9606.0   \n",
       "3     PP125278.1        Rhinovirus B              147712             9606.0   \n",
       "4     PP028461.1  Coxsackievirus A24               12089             9606.0   \n",
       "...          ...                 ...                 ...                ...   \n",
       "4550  AB769152.1  Coxsackievirus A24               12089             9606.0   \n",
       "4551  AB705311.1        Echovirus E6               12062             9606.0   \n",
       "4552  AB705310.1        Echovirus E6               12062             9606.0   \n",
       "4553  AB705309.1        Echovirus E6               12062             9606.0   \n",
       "4554  AB705308.1        Echovirus E6               12062             9606.0   \n",
       "\n",
       "     Completeness Isolate Collection date          Release date  \n",
       "0        COMPLETE                    2020  2024-01-24T00:00:00Z  \n",
       "1        COMPLETE                    2020  2024-01-24T00:00:00Z  \n",
       "2        COMPLETE              2021-06-07  2024-01-24T00:00:00Z  \n",
       "3        COMPLETE              2021-05-13  2024-01-24T00:00:00Z  \n",
       "4        COMPLETE                 2023-09  2024-01-11T00:00:00Z  \n",
       "...           ...                     ...                   ...  \n",
       "4550     COMPLETE                 2011-06  2012-12-05T00:00:00Z  \n",
       "4551     COMPLETE              2011-09-09  2012-09-14T00:00:00Z  \n",
       "4552     COMPLETE              2011-09-07  2012-09-14T00:00:00Z  \n",
       "4553     COMPLETE              2011-09-02  2012-09-14T00:00:00Z  \n",
       "4554     COMPLETE              2011-08-12  2012-09-14T00:00:00Z  \n",
       "\n",
       "[4555 rows x 7 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check that the right file was saved \n",
    "pd.read_csv(INPUT_PATH + output_filename, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2450be9a-928b-4851-8b03-4c9837b996e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_csv(INPUT_PATH + output_filename, sep='\\t')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
